pennylane/interfaces/jacobian_products.py

```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
+                tangent entry corresponds to the ``j`` th trainable parameter of the tape.
+
+        Returns:
+            ResultBatch, TensorLike: the results of the execution and the jacobian vector product
+
+        Side Effects:
+            caches newly computed results or jacobians if they were not already cached.
+
+        **Examples:**
+
+        For an instance of :class:`~.DeviceDerivatives` ``jpc``, we have:
+
+        >>> tape0 = qml.tape.QuantumScript([qml.RX(0.1, wires=0)], [qml.expval(qml.PauliZ(0))])
+        >>> tape1 = qml.tape.QuantumScript([qml.RY(0.2, wires=0)], [qml.expval(qml.PauliZ(0))])
+        >>> batch = (tape0, tape1)
+        >>> tangents0 = (1.5, )
+        >>> tangents1 = (2.0, )
+        >>> tangents = (tangents0, tangents1)
+        >>> results, jvps = jpc.execute_and_compute_jvp(batch, tangents)
+        >>> expected_results = (np.cos(0.1), np.cos(0.2))
+        >>> qml.math.allclose(results, expected_results)
+        True
+        >>> jvps
+        (array(-0.14975012), array(-0.39733866))
+        >>> expected_jvps = 1.5 * -np.sin(0.1), 2.0 * -np.sin(0.2)
+        >>> qml.math.allclose(jvps, expected_jvps)
+        True
+
+        While this method could support non-scalar parameters in theory, no implementation currently supports
+        jacobians with non-scalar parameters.
+
+        """
+        if tapes not in self._results_cache and tapes not in self._jacs_cache:
+            results, jacs = self._dev_execute_and_compute_derivatives(tapes)
+            self._results_cache[tapes] = results
+            self._jacs_cache[tapes] = jacs
+        else:
+            if tapes in self._results_cache:
+                if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+                    logger.debug("%s : Retrieving results from cache.", self)
+                results = self._results_cache[tapes]
+            else:
+                results = self._dev_execute(tapes)
+                self._results_cache[tapes] = results
+
+            if tapes in self._jacs_cache:
+                if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+                    logger.debug("%s : Retrieving jacobian from cache.", self)
+                jacs = self._jacs_cache[tapes]
+            else:
+                # Here the jac was not cached but the results were. This can not happen because results are never
+                # cached alone (note that in the else clause above computing only results, jac must already be present)
+                raise NotImplementedError(
+                    "No path to cache results without caching jac. This branch should not occur."
+                )
+
+        multi_measurements = (len(t.measurements) > 1 for t in tapes)
+        jvps = _compute_jvps(
+            jacs,
+            tangents,
+            multi_measurements,
+            has_partitioned_shots=tapes[0].shots.has_partitioned_shots,
+        )
+        return results, jvps
+
+    def compute_vjp(self, tapes, dy):
```

Comment:
See the next PR: https://github.com/PennyLaneAI/pennylane/pull/4637

That `DeviceJacobianProducts` class will be used when we want to use device provided vjps instead of device provided jacobians.

Reply:
Thank you I understand better :+1: